cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java

```diff
@@ -0,0 +1,155 @@
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.profile.UserProfile;
+
+import org.springframework.webflow.context.ExternalContextHolder;
+import org.springframework.webflow.context.servlet.ServletExternalContext;
+
+import java.security.GeneralSecurityException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.validation.constraints.NotNull;
+
+/**
+ * The Abstract Client Authentication Handler.
+ *
+ * This class is a generic handler for authentication delegated to an auth client, implementing the authentication
+ * method {@link AbstractPreAndPostProcessingAuthenticationHandler#doAuthentication(Credential)}. It uses the auth
+ * client and provides it with the client credentials in order to get the user profile returned by the provider for
+ * an authenticated user.
+ *
+ * @author Jerome Leleu
+ * @author Longze Chen
+ * @since 4.1.5
+ */
+@SuppressWarnings("unchecked")
+public abstract class AbstractClientAuthenticationHandler extends AbstractPreAndPostProcessingAuthenticationHandler {
+
+    /** The factory to create the principal type. **/
+    @NotNull
+    protected PrincipalFactory principalFactory = new DefaultPrincipalFactory();
+
+    /** The clients for authentication. */
+    @NotNull
+    private final Clients clients;
+
+    /**
+     * Instantiate a new {@link AbstractClientAuthenticationHandler} and define the clients.
+     *
+     * @param theClients the clients for authentication
+     */
+    public AbstractClientAuthenticationHandler(final Clients theClients) {
+        this.clients = theClients;
+    }
+
+    /**
+     * {@inheritDoc}
+     **/
+    @Override
+    public final boolean supports(final Credential credential) {
+        return credential != null && ClientCredential.class.isAssignableFrom(credential.getClass());
+    }
+
+    /**
+     * {@inheritDoc}
+     **/
+    @Override
+    protected HandlerResult doAuthentication(final Credential credential) throws GeneralSecurityException {
+
+        // Construct a specific client credential from a general credential
+        final ClientCredential clientCredentials = (ClientCredential) credential;
+        logger.debug("clientCredentials : {}", clientCredentials);
+
+        // Retrieve the client name found in the client credential
+        final Credentials credentials = clientCredentials.getCredentials();
+        final String clientName = credentials.getClientName();
+        logger.debug("clientName : {}", clientName);
+
+        // Attempt to load the client from defined clients
+        final Client<Credentials, UserProfile> client;
+        try {
+            client = this.clients.findClient(clientName);
+            logger.debug("client : {}", client);
+        } catch (final TechnicalException e) {
+            throw new DelegatedLoginException("Invalid client: " + clientName);
```

Comment:
Please add an error-level log when exception happens so that we can pinpoint the issue when login fails.